cmd/sink/filters/filters_test.go

```diff
@@ -142,40 +143,43 @@ func TestHandleUpdates(t *testing.T) {
 		},
 	}
 	var wg sync.WaitGroup
-	for i, testCase := range cases {
-		ch := make(chan watch.Event)
-		watcher := newMockWatcher(ch)
-		retryWatcher, err := toolsWatch.NewRetryWatcher("1", &cache.ListWatch{WatchFunc: watcher})
-		if err != nil {
-			assert.FailNow(t, "Could not create a watcher")
-		}
-		cm := corev1.ConfigMap{
-			TypeMeta: metav1.TypeMeta{
-				Kind:       "ConfigMap",
-				APIVersion: "v1",
-			},
-			ObjectMeta: metav1.ObjectMeta{
-				Name:            "sink-filters",
-				Namespace:       "kubearchive",
-				ResourceVersion: fmt.Sprintf("%d", i),
-			},
-			Data: testCase.data,
-		}
-		event := watch.Event{Type: watch.Modified, Object: &cm}
-		ch <- event
-		wg.Add(1)
-		go func(rw *toolsWatch.RetryWatcher, filter *Filters) {
-			defer wg.Done()
-			filter.handleUpdates(rw)
-		}(retryWatcher, f)
-		// we need to wait before calling retryWatcher.Stop to make sure that it sends a watch.Event first
-		time.Sleep(time.Millisecond * 500)
-		retryWatcher.Stop()
-		// wait until the filters update before checking the result of the update
-		wg.Wait()
-		assert.Equal(t, testCase.archiveSize, len(f.archive))
-		assert.Equal(t, testCase.deleteSize, len(f.delete))
-		assert.Equal(t, testCase.archiveOnDeleteSize, len(f.archiveOnDelete))
+	for i, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ch := make(chan watch.Event)
+			t.Cleanup(func() { close(ch) })
+			watcher := newMockWatcher(t, ch)
+			retryWatcher, err := toolsWatch.NewRetryWatcher("1", &cache.ListWatch{WatchFunc: watcher})
+			if err != nil {
+				assert.FailNow(t, "Could not create a watcher")
+			}
+			cm := corev1.ConfigMap{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "ConfigMap",
+					APIVersion: "v1",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:            "sink-filters",
+					Namespace:       "kubearchive",
+					ResourceVersion: fmt.Sprintf("%d", i),
+				},
+				Data: tt.data,
+			}
+			event := watch.Event{Type: watch.Modified, Object: &cm}
+			ch <- event
+			wg.Add(1)
+			go func(rw *toolsWatch.RetryWatcher, filter *Filters) {
+				defer wg.Done()
```

Comment:
In this case it does not matter, but for parallel tests, which we don't have yet, we should use `t.Cleanup` instead because defer does not guarantee order. See https://brandur.org/fragments/go-prefer-t-cleanup-with-parallel-subtests

Reply:
I'll make the change anyway to get myself in the right habit

Reply:
I _think_ this needs to remain a defer because `wg.Done` needs to be called when the `goroutine` that its in finishes not when the test it was called in completes. However, there might be something I'm missing

Reply:
but I removed the `defer` and made `wg.Done()` a direct call at the end of the function instead